// way to talk from server to client


syntax = "proto3";

package chatsystem;

// import "google/protobuf/empty.proto";

service ChatServer {
    
    rpc GetUser(User) returns (Status) {}

    rpc LogoutUser(User) returns (Status) {}

    rpc GetGroup(Group) returns (GroupDetails) {}

    rpc ExitGroup(Group) returns (Status) {}

    rpc GetMessages(Group) returns (stream Message) {}

    rpc PostMessage(Message) returns (Status) {}

    rpc HealthCheck(BlankMessage) returns (Status) {}
}

message BlankMessage {}

message Status {
    bool status = 1;
    string statusMessage = 2;
}

message Group {
    string group_id = 1;
    string user_id = 2;
}

message GroupDetails {
    string group_id = 1;
    repeated string users = 2;
    bool status = 3;
}

message User {
    string user_id = 1;
}

message Message {
    string group_id = 1;
    string user_id = 2;
    uint64 creation_time = 3;
    repeated string text = 4;
    string message_id = 5;
    map<string, int32> likes = 7;
    string message_type = 6;
}